pennylane/interfaces/jax_jit.py

```diff
@@ -312,3 +312,260 @@ def wrapped_exec_bwd(params, g):
+
+    Returns:
+        list[list[float]]: A nested list of tape results. Each element in
+        the returned list corresponds in order to the provided tapes.
+    """
+    # pylint: disable=unused-argument
+    if max_diff > 1:
+        raise InterfaceUnsupportedError("The JAX interface only supports first order derivatives.")
+
+    for tape in tapes:
+        # set the trainable parameters
+        params = tape.get_parameters(trainable_only=False)
+        tape.trainable_params = qml.math.get_trainable_indices(params)
+
+    parameters = tuple(list(t.get_parameters()) for t in tapes)
+
+    if gradient_fn is None:
+        return _execute_fwd_new(
+            parameters,
+            tapes=tapes,
+            device=device,
+            execute_fn=execute_fn,
+            gradient_kwargs=gradient_kwargs,
+            _n=_n,
+        )
+
+    return _execute_bwd_new(
+        parameters,
+        tapes=tapes,
+        device=device,
+        execute_fn=execute_fn,
+        gradient_fn=gradient_fn,
+        gradient_kwargs=gradient_kwargs,
+        _n=_n,
+    )
+
+
+def _numeric_type_to_dtype(numeric_type):
+    """Auxiliary function for converting from Python numeric types to JAX
+    dtypes based on the precision defined for the interface."""
+
+    single_precision = dtype is jnp.float32
+    if numeric_type is int:
+        return jnp.int32 if single_precision else jnp.int64
+
+    if numeric_type is float:
+        return jnp.float32 if single_precision else jnp.float64
+
+    # numeric_type is complex
+    return jnp.complex64 if single_precision else jnp.complex128
+
+
+def _extract_shape_dtype_structs(tapes, device):
+    """Auxiliary function for defining the jax.ShapeDtypeStruct objects given
+    the tapes and the device.
+
+    The host_callback.call function expects jax.ShapeDtypeStruct objects to
+    describe the output of the function call.
+    """
+    shape_dtypes = []
+
+    for t in tapes:
+        shape = t.shape(device)
+
+        tape_dtype = _numeric_type_to_dtype(t.numeric_type)
+        shape_and_dtype = jax.ShapeDtypeStruct(tuple(shape), tape_dtype)
+
+        shape_dtypes.append(shape_and_dtype)
+
+    return shape_dtypes
+
+
+def _execute_bwd_new(
+    params,
+    tapes=None,
+    device=None,
+    execute_fn=None,
+    gradient_fn=None,
+    gradient_kwargs=None,
+    _n=1,
+):  # pylint: disable=dangerous-default-value,unused-argument
+    total_params = np.sum([len(p) for p in params])
+
+    # Copy a given tape with operations and set parameters
+    def _copy_tape(t, a):
+        tc = t.copy(copy_operations=True)
+        tc.set_parameters(a)
+        return tc
+
+    @jax.custom_jvp
+    def execute_wrapper(params):
+        def wrapper(p):
+            """Compute the forward pass."""
+            new_tapes = [_copy_tape(t, a) for t, a in zip(tapes, p)]
+            with qml.tape.Unwrap(*new_tapes):
+                res, _ = execute_fn(new_tapes, **gradient_kwargs)

```

Comment:
Probably you have noticed this already, but the issue I described in #3218 arises because (I believe) this `execute_fn` here has an implicit device saved inside with a `prng_key`. Ideally, I think you should change the `execute_fn` to accept a device, and feed it as an explicit argument after having done something similar to `copy_tape` to replace eventual tracers (`_prng_key` in it).